### PR TITLE
Add is_null/is_not_null/any_not_null filter operators and Flag field support

### PR DESCRIPTION
**Pull request by shayp**
## Summary

Adds new filter operators to the JSON-based filter framework and fixes Flag field handling in featuremap VCF-to-Parquet conversion.

### 1. New filter operators (`filter_dataframe.py`)

Added three operators to the filter framework for annotation-based filtering:

- `is_null` — passes if the field is absent (null). Used for training exclusion: `{"field": "EXCLUDE_TRAINING", "op": "is_null"}`
- `is_not_null` — passes if the field is present. Used for inference inclusion on a single field.
- `any_not_null` — passes if at least one of multiple fields is present (OR logic). Used for inference inclusion: `{"op": "any_not_null", "fields": ["INCLUDE_INFERENCE", "PCAWG"]}`

These are unary operators (no `value` key needed). Validation updated to handle `any_not_null` using `fields` list instead of single `field`.

### 2. Flag field dtype fix (`featuremap_to_dataframe.py`)

- **Flag to pl.Utf8** instead of pl.Boolean: `bcftools query` outputs `1` for present flags and `.` for absent. Polars cannot parse `1` as Boolean (expects true/false). With Utf8, `.` becomes null via null_values and `1` stays as non-null string.
- **Flag cast passthrough**: Flag columns no longer get fill_null(False).cast(Boolean) — just pass through as-is.
- **Row count logging**: Log final parquet row count on completion.

### 3. Tests (`test_annotation_filtering.py`)

11 unit tests covering: is_null, is_not_null, any_not_null, combined filters, edge cases.

## Test plan
- [x] Unit tests pass (11 tests)
- [x] Pre-commit checks pass
- [x] Omics run 9275938 (CL10377) completed successfully with Flag annotations + JSON filters
- [x] Omics run 2885148 (CL10377) verified annotation filters active in logs

---

> [!NOTE]
> **Medium Risk**
> Extends the JSON filter language and changes how VCF `Flag` INFO/FORMAT fields are typed during VCF→Parquet conversion, which can alter filtering outcomes and downstream data expectations.
> 
> **Overview**
> Adds three new unary filter operators to the JSON filtering framework—`is_null`, `is_not_null`, and multi-field `any_not_null`—including updated validation and naming logic to support `fields`-based rules.
> 
> Fixes VCF `Flag` field ingestion in `featuremap_to_dataframe` by treating flags as `Utf8` (so bcftools `1`/`.` values work with null-based filters) and stops casting flags to boolean; also logs the final Parquet row count on completion.
> 
> Introduces `prepare_annotation_vcfs` CLI to generate/augment filter JSONs for annotation-driven exclusion/inclusion (coverage threshold updates, injected `is_null` rules, and `any_not_null` inference filters), and adds unit tests covering the new operators and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5e8ac0d32068cfe70cdead5f7f5db521dfad2dad. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by shayp** [comment]
@GatKrieger  i made changes of including the filtering to be part of the json an ran a test
outptut can be found here
https://proxy.ultimagen.dev/s3/ultimagen-pipelines-380827583499-us-east-1-outputs/SingleReadSNV/_data_Runs_perchik_tapas_downsample_srsnv_jsons_CL10377_dnn_multi_vcf_1fold/2885148/out/

**Comment by shayp** [comment]
@GatKrieger  here you can find the latest outputs with fixed vcf output
https://us-east-1.console.aws.amazon.com/omics/home?region=us-east-1#/runs/9275938
i think we are now ready for pr

File: src/featuremap/ugbio_featuremap/filter_dataframe.py [hunk 1]
```diff
@@ -84,34 +84,44 @@
...
-
-    # Check type
     valid_types = {TYPE_QUALITY, TYPE_REGION, TYPE_MAPPING, TYPE_LABEL}
     if rule[KEY_TYPE] not in valid_types:
         raise ValueError(f"Filter {index} has invalid type '{rule[KEY_TYPE]}'. Must be one of: {valid_types}")
 
-    # Check value/value_field
-    if KEY_VALUE not in rule and KEY_VALUES not in rule and KEY_VALUE_FIELD not in rule:
-        raise ValueError(f"Filter {index} must have either '{KEY_VALUE}', '{KEY_VALUES}', or '{KEY_VALUE_FIELD}'")
+    if rule[KEY_OP] == "any_not_null":
```
Comment (kenissur) [inline src/featuremap/ugbio_featuremap/filter_dataframe.py]: use constants instead of "any_not_null" ,"fields" etc.

**Comment by GatKrieger** [comment]
> @GatKrieger here you can find the latest outputs with fixed vcf output https://us-east-1.console.aws.amazon.com/omics/home?region=us-east-1#/runs/9275938 i think we are now ready for pr

Thanks. I'm looking at the dnn_featuremap_vcf and see most of the entries don't have INCLUDE_INFERENCE
was that intended?
this is the file: s3://ultimagen-pipelines-380827583499-us-east-1-outputs/SingleReadSNV/_data_Runs_perchik_tapas_downsample_srsnv_jsons_CL10377_dnn_multi_vcf_1fold/9275938/out/dnn_featuremap_vcf/402572-CL10377-dnn-multi-vcf.dnn_annotated.featuremap.vcf.gz

**Comment by shayp** [comment]
> > @GatKrieger here you can find the latest outputs with fixed vcf output https://us-east-1.console.aws.amazon.com/omics/home?region=us-east-1#/runs/9275938 i think we are now ready for pr
> 
> Thanks. I'm looking at the dnn_featuremap_vcf and see most of the entries don't have INCLUDE_INFERENCE was that intended? this is the file: s3://ultimagen-pipelines-380827583499-us-east-1-outputs/SingleReadSNV/_data_Runs_perchik_tapas_downsample_srsnv_jsons_CL10377_dnn_multi_vcf_1fold/9275938/out/dnn_featuremap_vcf/402572-CL10377-dnn-multi-vcf.dnn_annotated.featuremap.vcf.gz

I keep all the reads that snvfind outputs. i just run inference on those with the INCLUDE_INFERENCE/ PCAWG and for those i add valid snvq and mqual

i think it make sense, do you want it to work different?

**Review by GatKrieger — APPROVED** [review]
I don't see any problem here, but please wait till the terra branch is approved before you merge this one

**Comment by shayp** [comment]
Addressing review comments:

**kenissur — use constants:** Done — defined `OP_IS_NULL`, `OP_IS_NOT_NULL`, `OP_ANY_NOT_NULL`, `KEY_FIELDS` constants and replaced string literals.

**Cursor Bot — KeyError for any_not_null:** Fixed — fallback name generation uses `KEY_FIELDS` when `KEY_FIELD` is absent.

**Cursor Bot — Empty fields list crashes:** Fixed — validation checks `len(rule[KEY_FIELDS]) > 0`.

**Cursor Bot — Flag dtype change broadly affects all Flag fields:** Intentional fix. `bcftools query` outputs `1`/`.` for Flags; Polars can't parse `1` as Boolean (expects `true`/`false`). The old `pl.Boolean` was already broken for any VCF with Flag fields (e.g. STR). Added explanatory comment.

**Cursor Bot — Annotation filter redundantly applied twice / Orphaned rsID keys:** Not applicable — both `_apply_annotation_filters` and `_strip_unwanted_info` were removed in later commits. The bot reviewed intermediate commits that no longer exist in the final diff.

**Comment by kenissur** [comment]
Agree with Gat, looks good assuming terra_pipeline runs look as expected
